### DISCUSSION TITLE
[Feature] Add OCR

### DISCUSSION THREAD

**Discussion by chriexpe** [immich-app/immich#3168] [Feature Request]
Edit: yay thank you @mertalev and especially @flipped-1121!

### The feature

OCR is the only thing that keeps bringing me back to Google Photos, as I have A TON of screenshots, memes and documents so it's really handy to find exactly what I want.

I've seen somewhere here someone using this repo for OCR [PaddlePaddle/PaddleOCR](https://github.com/PaddlePaddle/PaddleOCR), hopefully this might help you integrate into Immich!

Thank you guys for this amazing project!

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by klaus1k** [comment]
I really like this feature on GPhotos!
It seems to me that it already works with CLIP? If I search for text in my pictures in Immich it shows the pictures with this text in them.

**Reply by chriexpe** [reply]
I think that might be just a coincidence, in my case most of the time the results are random

**Reply by NicholasFlamy** [reply]
> I think that might be just a coincidence, in my case most of the time the results are random

It's CLIP which has quasi OCR capabilities. It can sometimes match text when you search. It works better with better models but even then it's still not true OCR.

**Reply by Fireflaker** [reply]
In my experience - CLIP OCR barely work at all. Screenshots with arbitrary text do not return in smart search.

**Reply by NicholasFlamy** [reply]
@Fireflaker I said sometimes. Also, try a better CLIP model.

**Comment by a45s67** [comment]
I agree with @chriexpe.

I have tons of meme pictures which are full of text. After trying the immich search engine for days, I found it is not easy to find out the image that has specific text contents.

> For instance, although CLIP learns a capable OCR system, when evaluated on handwritten digits from the MNIST dataset, zero-shot CLIP only achieves 88% accuracy, well below the 99.75% of humans on the dataset. 

According to the description in [CLIP](https://openai.com/research/clip), this model is not good at detecting the text in images. So I think if the search engine integrate with some OCR models, it might fill the lack of CLIP, and have a much perfect search results. There is a closed PR which could be a good implementation reference (https://github.com/immich-app/immich/pull/1200).

For example, here are some images I can't find when given corresponding text:
| | Images has corresponding text | Search result |
| --- | --- | --- |
|Text: My Confidence | <img width="1185" alt="image" src="https://github.com/immich-app/immich/assets/32732820/1dd9b144-42b9-486a-9c15-a8a4989e05d3"> |  <img width="1076" alt="image" src="https://github.com/immich-app/immich/assets/32732820/43b2b1e9-b51f-4c79-bb09-dad005240bf6"> |
| Text: XD | <img width="1129" alt="image" src="https://github.com/immich-app/immich/assets/32732820/575bf487-b3e9-46ce-8b19-d0e8a47191c7"> |  <img width="1221" alt="image" src="https://github.com/immich-app/immich/assets/32732820/9aed4e48-e8f5-4b9b-84c5-15978190bd02"> |

**Reply by NicholasFlamy** [reply]
Try a better model. Obviously this isn't a solution, but it will give better results.

**Comment by eagle470** [comment]
This is also a feature I use heavily. It would be supremely useful.

**Reply by eagle470** [reply]
I'll add clarity, I use it on my iPhone with apple photos. I have moved all images with text to immich similar to what was mentioned above.

I would do everything but that is currently not recommended by the developers.

**Comment by krebsman23** [comment]
Would be great to have that

**Comment by Bronya-Rayi** [comment]
Indeed, the OCR feature is very useful in cloud albums and I hope it will be added in subsequent updates!

**Comment by aisbergde** [comment]
For me OCR is major feature and part of my document archive workflow. My scanned documents are in G drive. In G drive, not in G Fotos. But screenshots are in G Photos and are scanned. So maybe I should even better think about how to store screenshots not in G Photos but in G drive.

Then OCR result in Photos would be much more clean. Anyway OCR is important.

**Reply by 2001Y** [reply]
私もこの意見に非常に同意します。
書類はファイルストレージにあるが、スクリーンショットは画像管理アプリにあるし、お店の看板や車のナンバーも写真管理アプリにあります。

OCRは機械学習による画像認識より軽量だと思うので是非実装してほしいです。

**Reply by sgofferj** [reply]
For that use case, maybe [paperless-ngx](https://docs.paperless-ngx.com/) would be a better alternative? You would have all "documents" in one app. It also comes with a mobile app which supports images being shared to.

**Comment by aviv926** [comment]
At least until it is officially added here is something you can get right now with CLIP model:

the CLIP model that exists in immich [supports OCR](https://openai.com/research/clip#:~:text=2.%20CLIP%20is%20flexible%20and%20general
) in a built-in way, **but not at a level of height accuracy**

Some models are trained on more data and some on less with different levels of accuracy.
CLIP does not set itself the primary goal of recognizing text in an image, but it is a type of option that exists, but it is certainly not something that can be relied on as OCR.

If you still want OCR (although I wouldn't recommend it for CLIP)
I would look for a balance between performance and results in the `SST2` database

A model like laion2b_s32b_b79k gets good performance in Average perf. on 38 datasets and also in SST2 so it can suit you.

For comparison the basic model that Immich is set to work with (ViT-B-32__openai) got only 0.5865 in SST2 compared to laion2b_s32b_b79k who got 0.6392 in SST2.
Although again, I wouldn't expect a high level of OCR from a CLIP model.

Important note: different models require different levels of computing power (FLOPs) you should check if this is something your system is capable of before choosing to change models.

Another note is that it is necessary to check whether the model exists in the Immich database
You can check it here:
https://huggingface.co/collections/immich-app/clip-654eaefb077425890874cd07

Sources regarding tests in SST2 can be found here:
https://github.com/mlfoundations/open_clip/blob/main/docs/openclip_results.csv#L84

**Reply by aisbergde** [reply]
If the user could switch between used models while searching, then one model could be used for normal search, another for OCR search, and other models for other reasons. That's why it would be fine, to store not only the results of tagging, but also the used model. This way assets could have different tag sets in parallel.

**Reply by aviv926** [reply]
I don't think this is something that will happen, multiple models mean a load on the system and that is not something that Immich is designed for.
In any case, I believe that sooner or later OCR will be a part of Immich. I believe that the difficult part is to find something of a very high level that can fit together with the other existing technologies in Immich.

**Reply by aisbergde** [reply]
why multiple models should generate a big load? Calculation happens only once per asset. If two models are used, then it just will double the time to apply. And when a model is used, than at a time only one model will be used. I see no issues to store metadata for multiple models, if the user wants this. If not, he can use only one model.

Therefore comparing several models would be much easier. It is just not possible in the moment. And changing the model means to lost all existing calculation.

**Reply by AngelaDMerkel** [reply]
I agree with @aisbergde and if performance is goal, it could be possible to run text recognition with a second model in sequence with the regular CLIP model. Users are already selecting model size and assessing themselves what level of performance impact they are willing to accept. Enabling a second model (optionally) doesn't seem conceptually much different.

**Reply by NicholasFlamy** [reply]
It would require a second label (in the db) on every image and more work to code it in. If you want a tool for comparing AI models, that is not what immich is for.

**Reply by aisbergde** [reply]
A lot of resources are used to apply models and to get the results. Switching models currently means loosing all this effort.

Another idea would be to not delete the existing results but to store them "on hold" for later reusing. Nothing needs to be changed in the existing implementation, but only a save and restore part could be added. This can be done on the database level:
- additional backup tables should be created
- procedures to backup and restore the tagging results.

These are simple database operations.

**Reply by NicholasFlamy** [reply]
You could just take regular backups. Like right now with the current immich you could take regular backups to accomplish some of these goals by restoring the backups to compare.

**Reply by aisbergde** [reply]
For me it would be much easier to add some tables and procedures to the existing database, because databases are my daily business. But I don't know what happens, when I add my own database objects (tables, views, procedures) in case of updates. Would my additional objects sustain? Normally they should.

The question: is it worth the effort, when these requirements are not really requested? If it is not supported via the GUI, the usage requires database knowledge. Currently the one existing result set is OK for me. I am happy with the face tagging and I don't use other tagging. For me much more important will be the possibility for manually tagging.

**Reply by AngelaDMerkel** [reply]
> If you want a tool for comparing AI models, that is not what immich is for.

I'm not skilled in this field at all, so please forgive me if I've explained poorly. Why would it not be possible to use an OCR tool like Tesseract and write any visible words to a little sidecar or somewhere in the database so that certain text strings can be searched and compared against transcribed content?

Would attempting to do so interfere with the ability to properly use the facial recognition models currently in use?

**Reply by NicholasFlamy** [reply]
Sorry @AngelaDMerkel, I should've @ the person I was talking to. You do realize you're asking for real OCR but @aisbergde seems to be asking for the ability to run multiple CLIP models at once so that you can switch while searching at least partly for the purpose of comparing models. I was attempting to address that comparing models can be done outside of immich and that the point of immich is a photo manager that has cool features like search using CLIP, not specifically a tool for comparing CLIP models.

**Comment by nodis** [comment]
me too

**Comment by ZtereoHYPE** [comment]
Voyage AI just released a [very promising model](https://blog.voyageai.com/2024/11/12/voyage-multimodal-3/) for multimodal (images and document screenshots) embeddings search. 
Perhaps this could be good enough to not require OCR to be implemented separately from the current CLIP search?

**Reply by blublub** [reply]
OH YES PLZ! - any kind of detection of writing on screens and signs would be awesome

**Reply by mertalev** [reply]
FYI that model isn't open source.

A model that can effectively understand text in images is nice, but I think the copy/paste-ability of the text is important besides the ability to search for it. Embedding models can somewhat do the latter and will get better at it over time as new models are released, but the former is a gap that can only be filled through a different kind of model.

**Reply by ctrlshiftm129** [reply]
> FYI that model isn't open source.
> 
> A model that can effectively understand text in images is nice, but I think the copy/paste-ability of the text is important besides the ability to search for it. Embedding models can somewhat do the latter and will get better at it over time as new models are released, but the former is a gap that can only be filled through a different kind of model.

Is OCR on the roadmap or will it be far away?

**Reply by mertalev** [reply]
I can say that I don't intend to work on it within the next 2 months. It's possible for it to be added before the stable release depending on timing, but more likely that it will come after.

**Reply by NicholasFlamy** [reply]
> what's wrong with tesseract, other than not being the shiny new thing?

I don't think anyone said there was anything wrong with it. You could phrase your question as: "Is tesseract a good option?"

**Comment by gcs8** [comment]
I am here to say I will 1000% throw down the $100 server license if I can search my memes/screenshots text/subtitles. That's part of why I am paying for gsuite.

**Reply by ZJahon** [reply]
I agree with you.Accurate ocr is a very important feature in daily life.

**Comment by QAQDE** [comment]
Need this function as well

**Comment by rekcodocker** [comment]
Second (third / forth ... ) this.  This would be a useful feature indeed.

**Comment by Aevaris** [comment]
I look forward to this feature. I tried implementing paperless NGX recently and was frankly disappointed. My main use cases are images (memes, pictures of whiteboards, pictures of writing/plaques from museums, etc) rather than actual PDFs which paperless didnt do a good job with. If/when Immich adds this, it will be huge.

**Reply by mertalev** [reply]
> unfortunately, i doubt adding OCR to immich would do much better, as paperless uses Tesseract, which is to my knowledge the best OCR engine.

This is definitely not true. There are plenty of models that outperform Tesseract.

**Reply by gcs8** [reply]
Also, getting something in a place nerds want to play with it will con them into putting in extra work to make it do the thing they want.

**Reply by Aevaris** [reply]
> > unfortunately, i doubt adding OCR to immich would do much better, as paperless uses Tesseract, which is to my knowledge the best OCR engine.
> 
> This is definitely not true. There are plenty of models that outperform Tesseract.

And even at a worst case, if it is as 'good' as Paperless, it would meet my expectations as I'd have everything under 1 roof (dont need to maintain 2 apps, have users go to 2 apps, avoids duplication of source files). The UI and general experience of Immich far exceeds that of paperless (for me).

If the model performs better than Paperless, then that is a bonus and it would exceed my expectations.

I expect with time (e.g. if i think 5 years from now), all models will improve so I'm not terribly concerned about the current accuracy.

**Comment by Fireflaker** [comment]
TLDR: Yes I want OCR too

I am not looking for high accuracy OCR on pages full of text, but OCR will indefinitely come in handy for screenshots - and I believe it will significantly improve search accuracy. Usually on google photo- when smart search can't get specific enough, I would search for some random text that I remembered in that photo- and it usually pulls the right photo right up.

**Comment by Martin22** [comment]
Just a small idea - whether it might be better to implement VLM directly instead of OCR, but I understand that it's more computationally demanding than combining OCR and CLIP.

**Reply by rekcodocker** [reply]
While it would be nice I would be against it.

Two reasons. 

1. OCR has value on its own. Even just transcribing text in images will make them more useful and easier to find.

2. This is a big scope-creep. Textual description of images is new technology compared to OCR.
There is possibly no free model that is stable enough and also easy to integrate.
This requires more research, more work, not to mention large amounts of data and huge requirements for processing.
OCR on the other hand, is mature technology and available in open source. 

I am not saying it is impossible, or that it is a bad idea.  I am saying that scope creep has a proven track record of killing projects.

**Reply by Martin22** [reply]
I have tried many open-source OCR solutions myself, but unfortunately, the accuracy heavily depends on the chosen language model. I can send an image where Google correctly recognizes the text and makes it searchable, but with open-source OCR, I always end up with a garbled result. And yet, it's such a simple text, no diacritics, all uppercase letters.

Just to mention a few, I tried Tesseract, EasyOCR, Kraken OCR, MMOCR, and maybe some others I don't remember. I haven’t tested VLM locally, though. There are many open-source VLM models, but the smaller ones that can run efficiently on a CPU and provide fast results are likely not as accurate.

**Reply by Martin22** [reply]
I actually haven't tried PaddleOCR yet, which is mentioned right at the beginning, but it managed to recognize the text from my image flawlessly, well, if I don't take the issue with spacing into account. Unfortunately, Tesseract distorted the text from the image.

**Reply by NicholasFlamy** [reply]
> Just a small idea - whether it might be better to implement VLM directly instead of OCR, but I understand that it's more computationally demanding than combining OCR and CLIP.

VLM likely wouldn't accomplish the things most people would want from OCR. CLIP already gives you images based on words/descriptions, OCR is intended for extracting text from images (often documents) and there's been talks of further enhancing search with text from OCR.

I think VLM should have its own feature request. It might end up being closed as not planned but that's not my call.

Also, I think someone discussed the possibility of generating words and tagging images if the words match to a threshold with the CLIP embeddings.

**Comment by nyakang** [comment]
https://github.com/flipped-1121/immich/tree/ocr

I was eager to try OCR in Immich, so I built a local integration rather than wait for the official roadmap. The initial prototype supported both PaddleOCR and Tesseract, but Tesseract’s accuracy on non-English text was disappointing and it added extra maintenance overhead, so I’ve dropped it. **PaddleOCR**—consistently the stronger performer in my tests—is now the sole engine.

Because Immich is a sizable, sophisticated codebase, maintaining this feature alone long-term will be tough. If the community is interested, I’d welcome testers, feedback, and code contributions. I’ll keep the branch alive as long as there’s momentum, and once things are stable I’ll open a **PR** upstream. Keen to collaborate with anyone who’d like to help!

![image](https://github.com/user-attachments/assets/f6ec1742-c1a5-450f-884f-2c2a6ed0fc13)

https://github.com/user-attachments/assets/d61cd4ed-57fa-4518-abee-d0cb577c31a4

https://github.com/user-attachments/assets/4ed37925-a068-42be-abbf-cac0be2b97f7

**Reply by banjuer** [reply]
great job bro，i will try it

**Reply by banjuer** [reply]
咋用啊，有操作手册吗？

**Reply by mertalev** [reply]
Great work! Implementing a feature that spans different areas of the codebase is impressive. But this also means it will likely be difficult to maintain, and from a quick look it would need considerable changes to be mergeable upstream. If you want this to be merged into Immich, I recommend making a draft PR to make it easier to collaborate and review.

**Reply by thorgrin** [reply]
This is great, I really hope that this feature becomes mainstream. I frequently search for pictures containing particular pieces of text and have to resolve back to Samsung on device search or Google Photos (which I'm trying hard to migrate from).

**Reply by Aevaris** [reply]
This is fantastic. 

My personal experience with Tesseract (through a short stint with PaperlessNGX) was disappointing, so I am glad to hear another OCR model is out there that performs better. 

Really hope this or something similar goes mainstream with Immich. It'd be huge to be able to search for documents or memes and one of the last remaining things Google and OneDrive still do better.

**Reply by elmocito** [reply]
This is huge.
I so heavily want this 🤩

**Reply by nyakang** [reply]
> 咋用啊，有操作手册吗？


I’ve opened a draft PR. If you’d like to test it now, you can clone the branch and build it locally. Otherwise, I can carve out time to create a Docker container image of the service and share it.

**Reply by aliut** [reply]
Great Work, can you please share the docker images?, I am not able to build your repo for some reason, gives me quite alot of errors on ocr branch.

**Reply by ghost** [reply]
Awesome!

Have few questions:

1) I'm on weak machine with Immich and the ML is done by the [Remote ML](https://immich.app/docs/guides/remote-machine-learning).
Is this change compatible with it? How much RAM each model consume?
2) for install the build of docker image is needed locally and with that done just changing `docker-compose` for local build?
3) in case of need to switch back to main - what will happen to the files outside of containers (the state of OCR ML in database)?

**Reply by si9ma** [reply]
I tried it out and it feels pretty good. Overall it's usable!

**Reply by elmocito** [reply]
> I tried it out and it feels pretty good. Overall it's usable!

How? 😩

**Reply by nyakang** [reply]
https://github.com/flipped-1121?tab=packages&repo_name=immich

**Reply by dodysw3** [reply]
With Deepseek OCR model released, which seems to be the most powerful OCR to date, maybe consider adding support to this model?

**Comment by WuSiYu** [comment]
+1, I think add the OCR function is meaningful. It's not conflict with CLIP. And apps like Apple Photo has OCR and fuzzy search at same time too. It's extremely useful when searching photos about text material or sign.

**Comment by stefan1983** [comment]
+1 OCR would be extremely helpful

**Comment by Waffled-II** [comment]
Sad this hasn’t been really implemented yet. This is one of the few big features google photos has over this.

**Reply by Qhilm** [reply]
PR: https://github.com/immich-app/immich/pull/18836

Seems the mobile app code still needs some tweaking.

**Reply by reynold-lariza** [reply]
Yeah, I just discovered this just now, after a full night of migrating all my iCloud photos to it... XD

**Reply by m41denx** [reply]
I tried merging pr with latest release today and I have bad news. A LOT has changed in immich codebase since last commit in this pull request, especially search apis and some things in mobile app too. So I don't think we'll get OCR anytime soon

**Reply by elmocito** [reply]
Nooooooo 😭😭😭😭😭

**Reply by nyakang** [reply]
Thanks for the continued interest. Current status: this PR remains a draft with several items left to complete. 
Due to my schedule, I can’t commit to a steady cadence right now, but I will continue progressing on it as time permits.

**Reply by elmocito** [reply]
If there is anything (non-DEV) I can help you with. Please tell me.

**Comment by nyakang** [comment]
The OCR feature has been **merged** into the main branch, and a release can be expected.

**Reply by Qhilm** [reply]
this. is. amazing. Thank your SO much for this.

**Reply by icepigeon-dev** [reply]
Thank you so much

**Reply by elmocito** [reply]
Awesome news!! 🥳

**Reply by leepfrog-ger** [reply]
Thank you for your work, awesome!

**Reply by andrast0th** [reply]
Thank you ❤️

**Reply by kiber-io** [reply]
Does the OCR model used work only with English? It doesn’t detect anything in other languages :(

**Reply by elmocito** [reply]
Same here. 😭

**Reply by Blaskyy** [reply]
> Does the OCR model used work only with English? It doesn’t detect anything in other languages :(

Works with Chinese for me.

**Reply by kiber-io** [reply]
> Works with Chinese for me.

In my case, the Russian text doesn’t get searched.

**Reply by elmocito** [reply]
Any chance to add German?

**Reply by m41denx** [reply]
Looks like Paddle OCR v5 has optimized variants for non-english/chinese languages: https://www.paddleocr.ai/main/en/version3.x/algorithm/PP-OCRv5/PP-OCRv5_multi_languages.html

is there any chance it can these variants can be used over current `ch_PP-OCRv5` model?

**Reply by mertalev** [reply]
We can look into possibly supporting other variants. It seems feasible to expose it as a setting from a quick look.

**Reply by elmocito** [reply]
The German results look pretty promising. A language multi selector would be pretty nice.

**Reply by Qhilm** [reply]
Where is the language setting? I must be blind.

**Reply by chriexpe** [reply]
There is no language setting for now, the model being used only supports Chinese and English, but if your language has Latin alphabet it does work too.

**Reply by NicholasFlamy** [reply]
If y'all want a language setting, please go ahead and make a new feature request and stop messaging in this thread.
